### DISCUSSION TITLE
Running MAUI in Android

### DISCUSSION THREAD

**Discussion by Mamik153** [oqtane/oqtane.framework#2367] [Q&A]
I am trying to run MAUI project in android emulator but I am getting an issue saying appsettings.json is not found

![image](https://user-images.githubusercontent.com/30488855/184869722-efa62b44-403e-4213-a2c8-daffb3bc8b35.png)

![image](https://user-images.githubusercontent.com/30488855/184870089-dd3ef71d-6deb-4aeb-be7a-26fd76d242ed.png)

Any idea why this is happening? Was there an appsettings.json file that was supposed to be in Oqtane.Maui project which was omitted by mistake?

Also I am not getting option to run this on Windows platform. Any ideas??
![image](https://user-images.githubusercontent.com/30488855/184878982-68e52769-b5e9-496f-9601-224a4bc7dd90.png)

P.S. I loaded Oqtane.Client and Oqtane.Shared inside Oqtane.Maui sln which might be wrong on my part

**Comment by sbwalker** [comment]
#2369 deals with the appsettings.json issue. All of the Blazor Hybrid examples I have seen (such as the Podcast reference sample app from Microsoft) use hardcoding to define the Url for the server-side services. I am not a fan of hardcoding - and it definitely will not work in the real world - so I included a solution in Oqtane.Maui (not sure yet if its the right solution). Maui has a FileService abstraction service which works across device platforms. I used the FileSystem.Current.AppDataDirectory property to reference a custom appsettings.json file where the Url could be specified. On my system when running on Windows the folder is: 

C:\Users\Shaun.Walker\AppData\Local\Packages\0E29FC31-1B83-48ED-B6E0-9F3C67B775D4_9zz4h110yvjzm\LocalState\

And my appsettings.json file contains:

{
  "Url": "http://localhost:44357"
}

(note that if an appsettings is not used the app still uses a hardcoded Url of http://localhost:44357)

If the Windows option is not available to you then I assume you have not configured your environment properly for Maui. I am not sure why you loaded Oqtane.Client or Oqtane.Shared in the Maui solution - referencing the DLLs should work just fine.

**Reply by Mamik153** [reply]
Thanks for this. 

Initially I added the projects because it showed errors while building saying that Oqtane.Shared was not found (or something along this way). I actually added them then just for the sake of it, I knew I had to remove them later. 😅


But now I am have different issue. I am getting a "System.Net.WebException: Cleartext HTTP traffic to localhost not permitted".
I changed MainApplication.cs in Android to:

```
#if DEBUG
[Application(UsesCleartextTraffic = true)] 
#else                                       
[Application]                            
#endif
```
But then I am getting this error: System.Net.WebException: Failed to connect to localhost/127.0.0.1:44357 and I am trying to change this IP value. I think 127.0 refers to Android Device IP and the code is trying to find Server running on the Android device while it is running on Desktop


P.S. I think you are doing a bang on job and thanks for answering to queries 👍

**Reply by sbwalker** [reply]
When you get the error "Failed to connect to localhost/127.0.0.1:44357" is your Oqtane.Server project running? In my earlier instructions about MAUI I had indicated that I added Oqtane.Server to the Maui solution and then configure the solution to have 2 startup projects. The Server needs to be running in order for the client to connect to it. 

https://github.com/oqtane/oqtane.framework/discussions/2347#discussioncomment-3379298

**Reply by Mamik153** [reply]
Yes I followed the steps you mentioned in[ #2347 (comment)](https://github.com/oqtane/oqtane.framework/discussions/2347#discussioncomment-3379298) . This issue is happening when both the server and MAUI App is running. I even tried to run server separately and then run Oqtane.Maui on "Windows Machine" then I am getting a different error which says Oqtane.Maui.appxrecipe (something like this) cannot be found in bin/debug/net6.0-android.

I am at my wit's end here sorry in advance if this seems a bit childish. 

P.S. This is the first time I have dived in .NET let alone Blazor and Oqtane

**Comment by SunSecond** [comment]
maui app works good on Windows Machine but it seemed crash on Android.

**Comment by SunSecond** [comment]
![windows](https://user-images.githubusercontent.com/109659990/185363368-6870f206-81d9-473e-a30e-8087ab31b0c9.png)
![android](https://user-images.githubusercontent.com/109659990/185363386-27cb5f7d-671f-4ff9-8383-33dab5247bd1.png)

**Reply by sbwalker** [reply]
@SunSecond when you get this error can you hit F12 to load the browser dev tools and look in the console tab to see the actual error messages which are logged? I would be interested to see the errors in your environment.

**Reply by SunSecond** [reply]
![error](https://user-images.githubusercontent.com/109659990/186354796-01d01855-3388-4409-9638-27d597576454.png)
![error1](https://user-images.githubusercontent.com/109659990/186356664-9804202f-1cb5-4584-8c73-41cbc48263ce.png)

no any error messages  display when I hit F12,it seems that the maui does not call the  server API at all. maybe we need to build an oqtane site on IIS based on the IP address to test maui andriod.

**Reply by sbwalker** [reply]
You are using the code from the Dev branch without any modifications, correct? Can you also share a screen shot of your project startup configuration from Visual Studio - I do not understand why your server is not running if you configured it as a startup project.

**Reply by SunSecond** [reply]
I just downloaded the latest dev branch and built all project,web and windows works good,but when I run manu project after server project start on  windows command ,the error occurred as following below:
![r2](https://user-images.githubusercontent.com/109659990/186435511-50fcf2bc-2f88-4768-8359-c022839da815.png)
![r3](https://user-images.githubusercontent.com/109659990/186435522-7c500c0e-7a1b-47ce-ba8c-ab99eb9ee717.png)
![r1](https://user-images.githubusercontent.com/109659990/186435526-ff2c5e3a-8cd8-4cc1-96f4-17a39cd11ebc.png)

**Reply by SunSecond** [reply]
I installed all lost assembly and rebuild maui project,then run it as following screenshots:
![r7](https://user-images.githubusercontent.com/109659990/186440750-36435315-4043-442d-8063-747b56b0758e.png)
![r4](https://user-images.githubusercontent.com/109659990/186440711-5a2ac817-aa99-435c-bb13-fdc5fe5cf020.png)
![r5](https://user-images.githubusercontent.com/109659990/186440722-53c6090c-cceb-49de-982b-b2160bc53600.png)
![r6](https://user-images.githubusercontent.com/109659990/186440738-88bccf73-d055-42c4-96d7-b499111ce7a8.png)

**Reply by SunSecond** [reply]
Got same error when I configure the Maui solution to have 2 startup projects.
![r8](https://user-images.githubusercontent.com/109659990/186455414-3e327b3f-39fc-4192-8fa9-a4d01354968b.png)

**Reply by sbwalker** [reply]
The errors seems to indicate that assembly references are missing - however it seems very strange that it runs fine as a Windows Desktop app but not as an Android app.

**Comment by sbwalker** [comment]
thank you for letting me know... I will need to investigate

**Reply by sbwalker** [reply]
I am still exploring the Android issue... it has taken me some time to get my environment configured with hardware acceleration. I believe the problem is that the assembly reference to Oqtane.Client is not allowing the Maui project to pick up all of the necessary dependencies. However it is very strange that it works on Windows desktop but not Maui. This issue needs to be resolved before 3.2 is released.

**Comment by sbwalker** [comment]
Ok, so after much trial and error I have a better understanding of this area...

When running Blazor Hybrid on Windows Desktop it is important to recognize that the client and server are running on the same machine. This means that all of the networking configuration is already wired up for you ie. localhost will be resolved automatically, etc.. However when you try to run Blazor Hybrid on a different device - whether that is a Mac, mobile device, or emulator it is a completely independent environment, so half the battle is figuring out the network configuration which will enable the client and server to communicate with one another. I suspect this was the problem you experienced when running Oqtane on the Android emulator.

The following link is helpful in setting up your development environment:
https://docs.microsoft.com/en-us/aspnet/core/blazor/hybrid/tutorials/maui?view=aspnetcore-6.0

This link describes how to enable hardware acceleration which I would highly recommend when using the Android Emulator as otherwise it is extremely slow:
https://docs.microsoft.com/en-us/dotnet/maui/android/emulator/hardware-acceleration

Another link on how to use Visual Studio for debugging on Android Emulator:
https://docs.microsoft.com/en-us/dotnet/maui/android/emulator/debug-on-emulator

I got the environment installed and then began working through the various issues I encountered (which were displayed in the Visual Studio console):

"cleartext HTTP traffic to localhost not permitted"

This error is related to default emulator settings not allowing non-secure communication with the server (Oqtane.Server is running on HTTP in my environment - not HTTPS). The recommended solution seemed to be to add android:usesCleartextTraffic="true" to the AndroidManifest.xml.... but this produced the next error:

"Failed to connect to localhost/127.0.0.1:44357"

That is a strange looking Url. It turns out the Android Emulator is an independent device so when you reference localhost it thinks you are referring to the localhost of the emulator itself. So this approach does not work. The Android Emulator has a convention of using 10.0.2.2 instead which will resolve to your dev machine's localhost.

The following link is very helpful with the required configurations:
https://docs.microsoft.com/en-us/xamarin/cross-platform/deploy-test/connect-to-local-web-services

I added a Platforms/Android/Resources/xml/network_security_config.xml:

```
<?xml version="1.0" encoding="utf-8"?>
<network-security-config>
  <domain-config cleartextTrafficPermitted="true">
    <domain includeSubdomains="true">10.0.2.2</domain>
  </domain-config>
</network-security-config>
```

And I updated Platforms/Android/AndroidManifest.xml to include:

android:networkSecurityConfig="@xml/network_security_config"

And then on to the next error:

Request: http://10.0.2.2:44357/api/Installation/installed/
BadRequest Bad Request
Object reference not set to an instance of an object.

At least the Url was correct this time. This error is related to IIS Express using default bindings which point explicitly at localhost... which effectively disables remote access. And since the Android Emulator is a remote client it is unable to reach the server.

The following blog was helpful in configuring IIS Express:
https://weblog.west-wind.com/posts/2016/sep/28/external-network-access-to-kestrel-and-iis-express-in-aspnet-core

I modified /oqtane.framework/.vs/Oqtane.Maui/config/applicationhost.config to change binding protocol from "*:44357:localhost" to "*:44357:*"

```
      <site name="Oqtane.Server" id="2">
        <application path="/" applicationPool="Oqtane.Server AppPool">
          <virtualDirectory path="/" physicalPath="C:\Users\Shaun.Walker\source\repos\sbwalker\oqtane.framework\Oqtane.Server" />
        </application>
        <bindings>
          <binding protocol="http" bindingInformation="*:44357:*" />
        </bindings>
      </site>
```

And I ran some networking commands (as Administrator) to open up communication:

netsh http add urlacl url=http://*:44357/ user=Interactive listen=yes
netsh advfirewall firewall add rule name="Http Port 44357" dir=in action=allow protocol=TCP localport=44357

Now the Android Emulator could reach the server.. however my multiple startup project configuration for my Visual Studio solution had somehow been lost so the initial call from the Android Emulator was still failing because Oqtane.Server was not running. So I re-added  the multiple startup project configuration for my solution so that Oqtane.Server would startup automatically and be running when the Android Emulator tried to connect. 

And then I got a new error... this time it was an Oqtane error because I had not registered the new site alias. So I added an entry for 10.0.2.2:44357 to the Alias table in my database (setting IsDefault=True so it would not redirect).

And voila... Oqtane was running in the Android Emulator. I was able to login and navigate to all of the administrative areas. 

![image](https://user-images.githubusercontent.com/4840590/187765878-7b6316e6-4ec6-4e26-bc7b-1e7aa0047bd3.png)

You will notice from the screen shot that there is a still a problem. Some of the static resources which are supposed to be loaded from the server are failing (ie. CSS and images). The console is showing the following error:

Mixed Content: The page at 'https://0.0.0.0/admin/site' was loaded over HTTPS, but requested an insecure image 'http://10.0.2.2:44357/files/logo-white.png'. This request has been blocked; the content must be served over HTTPS.

This is happening because apparently the Android Emulator is starting up using HTTPS - whereas the server is using HTTP. The HttpClient instance in Oqtane.Maui is configured to call API services on HTTP - so those are functioning just fine. It is just the static references in the HTML markup which are failing (as those are not handled by HttpClient - they are handled by the Web View itself). 

I am still looking for a solution to this problem - hoping there is a configuration option for MixedContentMode.

Changes included in #2389

**Reply by leigh-pointer** [reply]
This awesome @sbwalker

**Reply by SunSecond** [reply]
great exploring @sbwalker

**Reply by Sen-Gupta** [reply]
@sbwalker The weekend is going to be fun. ;-)

**Comment by sbwalker** [comment]
After further investigation, there is no MixedContentMode supported in a BlazorWebView. So this means that the only real option is to avoid mixed content scenarios - which means that the Oqtane.Server would need to be running on HTTPS so that it is consistent with the Android Emulator expectation. 

Running Oqtane.Server on HTTPS is simple... you can modify \Oqtane.Server\Properties\launchSettings.json and specify a value for sslPort (currently it is set to zero and you will need to choose something other than 44357). Let's use port# 44358 (IIS Express supports 44300-44399 by default). This port will need to be registered as an Alias in the database (ie. localhost:44358). Once these changes are made you can run Oqtane.Server and it will startup on HTTPS (https://localhost:44358).

To run the Android Emulator you will need to repeat some of the earlier steps to configure the environment to use the new secure port#. The default base url is specified in MauiProgram.cs:

    static string url = (DeviceInfo.Platform == DevicePlatform.Android) ? "https://10.0.2.2:44358" : "https://localhost:44358";

And we need to specify the binding in /oqtane.framework/.vs/Oqtane.Maui/config/applicationhost.config:

```
      <site name="Oqtane.Server" id="2">
        <application path="/" applicationPool="Oqtane.Server AppPool">
          <virtualDirectory path="/" physicalPath="C:\Users\Shaun.Walker\source\repos\sbwalker\oqtane.framework\Oqtane.Server" />
        </application>
        <bindings>
          <binding protocol="https" bindingInformation="*:44358:*" />
        </bindings>
      </site>
```
and run the commands (as Administrator) to open up communication:

netsh http add urlacl url=https://*:44358/ user=Interactive listen=yes
netsh advfirewall firewall add rule name="Http Port 44358" dir=in action=allow protocol=TCP localport=44358

And add an Alias to the database for the Android Emulator IP Address and new port# "10.0.2.2:44358" 

But... no luck... the project throws a Visual Studio error on startup "Invalid URI: The hostname could not be parsed.". After fighting with this, I disabled the Oqtane.Server project in the Oqtane.Maui solution and ran it in the standard Oqtane solution instead - where it ran without any issues. Then I ran Oqtane.Maui and it got further before it failed with a security issue:

"java.security.cert.CertPathValidatorException: Trust anchor for certification path not found."

This appears to be a certificate problem... which I am currently investigating. 

So far none of the issues encountered have anything to do with Oqtane itself - they are all related to environment configuration and limitations.

**Comment by sbwalker** [comment]
So running the Android Emulator locally using a secure connection to the server seems to be a known problem:

https://github.com/dotnet/maui/discussions/8131

The issue is that the SSL certificate is validated on the client using the certificate store in the emulator, which rejects personal certificates. There is a workaround for HttpClient API requests... but the requests for static resources such as images and CSS which reside on the server would still fail. So this is a dead end. 

The last remaining option is to connect the Android Emulator to a remote public server running Oqtane with a proper SSL certificate. Hopefully this will bypass all of the local development environment limitations... but even if it is successful, it represents a poor development experience in .NET MAUI :( Hopefully Microsoft can improve this in the future. 

Ideally, the simplest solution would be developing on HTTP locally - but in order for that to be viable there needs to be a way to startup the BlazorWebView on http:// rather than https:// ... or allow the BlazorWebView to render mixed content. Neither option is available currently.

**Comment by Sen-Gupta** [comment]
Host Oqtane on IIS with a binding and cert...
Everything should work afterwards...

**Reply by sbwalker** [reply]
Yes, that is essentially the same approach that I outlined above "connect to remote public server running Oqtane with a proper SSL certificate" - but use IIS on your own machine instead. This may be viable - however, the effort to get IIS configured on a local development machine is not trivial... so basically this does not align with the "low ceremony" goals for Oqtane (ie. download the code, load the solution and run). But I suppose mobile development is more complicated just by its very nature of coordinating with multiple devices, so perhaps this is fine for now.

**Reply by Sen-Gupta** [reply]
Just spin oqtane on docker, does docker go with low ceremony? 

We are running oqtane on docker, all that was required to define two volumes.

**Comment by sbwalker** [comment]
I experienced some difficulty connecting the Android Emulator to a remote public server running Oqtane. The problem was related to the fact that MAUI does not support an asynchronous startup method similar to async Task Main() which exists for Blazor WebAssembly. Oqtane needs to download assemblies and load them into the App Domain during startup - which requires an asynchronous HttpClient request. So I had to include some ugly logic for calling async code synchronously. The initial logic worked for Windows Desktop but not Android Emulation (it would cause an unrecoverable deadlock). I was finally able to get this resolved by using a different syntax:
 
```
            var zip = Task.Run(() => http.GetByteArrayAsync("/api/Installation/load")).GetAwaiter().GetResult();
```

This allowed the logic to execute properly without deadlock (in the long term it would still be preferable to have an async startup method but that will have to wait for a future MAUI release). Here is an example of the https://www.dnfprojects.com site running in the Android Emulator:

![image](https://user-images.githubusercontent.com/4840590/189375496-37fdc788-173e-499f-9cae-09dd50b77d8e.png)
